sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -375,4 +423,372 @@ public void checkDone() throws IllegalStateException {
       }
     }
   }
+
+  /**
+   * A splittable {@link DoFn} which executes an {@link UnboundedSource}.
+   *
+   * <p>We model the element as the original source and the restriction as a pair of the sub-source
+   * and its {@link CheckpointMark}. This allows us to split the sub-source over and over as long as
+   * the checkpoint mark is {@code null} or the {@link NoopCheckpointMark} since it does not
+   * maintain any state.
+   */
+  @UnboundedPerElement
+  static class UnboundedSourceAsSDFWrapperFn<OutputT, CheckpointT extends CheckpointMark>
+      extends DoFn<UnboundedSource<OutputT, CheckpointT>, ValueWithRecordId<OutputT>> {
+    private static final Logger LOG = LoggerFactory.getLogger(UnboundedSourceAsSDFWrapperFn.class);
+    private static final int DEFAULT_DESIRED_NUM_SPLITS = 20;
+    private static final int DEFAULT_BUNDLE_FINALIZATION_LIMIT_MINS = 10;
+    private final Coder<CheckpointT> restrictionCoder;
+
+    private UnboundedSourceAsSDFWrapperFn(Coder<CheckpointT> restrictionCoder) {
+      this.restrictionCoder = restrictionCoder;
+    }
+
+    @GetInitialRestriction
+    public KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT> initialRestriction(
+        @Element UnboundedSource<OutputT, CheckpointT> element) {
+      return KV.of(element, null);
+    }
+
+    @GetSize
+    public double getSize(
```

Comment:
Do we assume that `getSize` here is always called for calculating backlog in streaming? It is not correct for calculating  primary size(though primary size doesn't matter for now).

Reply:
getSize is used for the initial size and isn't used anywhere for progress/backlog reporting yet.

With #11472, this can be migrated to HasProgress on the restriction tracker.